test/library/packages/ConcurrentMap.skipif

```diff
@@ -0,0 +1,11 @@
+# AtomicObjects uses extern blocks
+CHPL_LLVM==none
```

Comment:
It seems like what's really needed here is something like "`chpl` was not built with LLVM enabled" which is arguably different than what `CHPL_LLVM` is set to, or inferred to be, at the time of testing?

Reply:
I thought it might be cool to add a `proc isExternBlockSupportEnabled() param`, but on second thought it might be cool to do something like:

```
AtomicObjects:3: CHPL_LLVM must not be none to use extern blocks
ConcurrentMap: 2: Used by module 'ConcurrentMap' here
```

Reply:
> CHPL_LLVM must not be none to use extern blocks

Once again, I don't think the constraint is "CHPL_LLVM can't be none."  Can't I do:

```
export CHPL_LLVM=system
make compiler
export CHPL_LLVM=none
chpl test/release/examples/test-uses-extern-block.chpl  # will work because even though CHPL_LLVM=none, `chpl` was built with LLVM enabled
```

and have things work fine?

Reply:
I should probably stop using the `CHPL_LLVM` variable as an example, because that's just reading from the `ENV`.

The only thing I care about is whether or not the current build of the compiler supports compiling extern blocks, and providing a way to detect that in module code. But even if we detect that and let the user write a nice error message, I'm not sure if that's enough because that extern block warning is probably emitted at parse time. So I was suggesting a chain of use statements as an alternative, but that probably doesn't work either.

Reply:
> I just want to think of ways to have the compiler issue error messages that are more informative rather than rely on documentation alone.

What's the behavior today?  And can the parse-time check that detects the extern block when LLVM isn't available simply be spruced up?

Reply:
It's emitted when we parse an extern block right now:

```cpp
  // Check that the compiler supports extern blocks
  // but skip these checks for chpldoc.
  if (fDocs == false) {
#ifdef HAVE_LLVM
    // Chapel was built with LLVM
    // Just bring up an error if extern blocks are disabled
    if (fAllowExternC == false)
      USR_FATAL(ret, "extern block syntax is turned off. Use "
                     "--extern-c flag to turn on.");
#else
    // If Chapel wasn't built with LLVM, we can't handle extern blocks
    USR_FATAL(ret, "Chapel must be built with llvm in order to "
                    "use the extern block syntax");
#endif
  }
```

I think we would have to go ahead and parse it and delay emitting errors until scope resolution or later.

I'm not sure how feasible/worth it is to do in the current compiler, but in `compiler/next` I think this should be pretty easy to do.

Reply:
On second thought, this might actually require something like the "parse-time-if". I'm not entirely sure/can't convince myself that this would work even if we delayed lowering the extern block.

Reply:
I'm not clear why we'd need to wait until param folding occurs, since use and import clauses aren't affected by param folding today (for better or worse).  At parse time we know the file we're currently parsing and must've known what module caused us to parse that module—though we may not maintain that information well enough today to be able to re-create the full use chain without more book-keeping (which, as you note, would start to get into the realm of "maybe we should just wait for the new compiler front-end for this").  But if I'm misremembering and we already do maintain a "we're parsing this module because of that module" information, we should be able to walk the information back out.